go-controller/pkg/node/controllers/upgrade/node_upgrade.go

```diff
@@ -6,231 +6,133 @@ import (
-	// has been synced at least once. Added as a member to the struct to allow
-	// injection for testing.
-	nodesSynced cache.InformerSynced
-
-	queue workqueue.RateLimitingInterface
-
-	// Holds the initially found version
-	initialTopoVersion int
+	client kubernetes.Interface
+	wf     factory.NodeWatchFactory
 }
 
 // NewController creates a new upgrade controller
-func NewController(client kubernetes.Interface) *upgradeController {
+func NewController(client kubernetes.Interface, wf factory.NodeWatchFactory) *upgradeController {
 	uc := &upgradeController{
 		client: client,
-		queue:  workqueue.NewNamedRateLimitingQueue(workqueue.DefaultControllerRateLimiter(), controllerName),
-	}
-
-	// note this will change in the future to control-plane:
-	// https://github.com/kubernetes/kubernetes/pull/95382
-	masterNode, err := labels.NewRequirement("node-role.kubernetes.io/master", selection.Exists, nil)
-	if err != nil {
-		klog.Fatalf("Unable to create labels.NewRequirement: %v", err)
-	}
-
-	labelSelector := labels.NewSelector()
-	labelSelector = labelSelector.Add(*masterNode)
-
-	uc.informerFactory = informers.NewSharedInformerFactoryWithOptions(client, 0,
-		informers.WithTweakListOptions(func(options *metav1.ListOptions) {
-			options.LabelSelector = labelSelector.String()
-		}))
-	nodeInformer := uc.informerFactory.Core().V1().Nodes()
-	klog.Info("Setting up event handlers for node upgrade")
-	nodeInformer.Informer().AddEventHandler(cache.ResourceEventHandlerFuncs{
-		AddFunc:    uc.onNodeAdd,
-		UpdateFunc: uc.onNodeUpdate,
-	})
-	uc.nodeLister = nodeInformer.Lister()
-	uc.nodesSynced = nodeInformer.Informer().HasSynced
-	uc.initialTopoVersion, err = uc.detectInitialVersion(labelSelector)
-	if err != nil {
-		klog.Fatalf("Unable to run initial version detection: %v", err)
+		wf:     wf,
 	}
 	return uc
 }
 
-// onNodeAdd queues the Node for processing.
-func (uc *upgradeController) onNodeAdd(obj interface{}) {
-	key, err := cache.MetaNamespaceKeyFunc(obj)
-	if err != nil {
-		utilruntime.HandleError(fmt.Errorf("couldn't get key for object %+v: %v", obj, err))
-		return
-	}
-	klog.V(4).Infof("Adding node %s", key)
-	uc.queue.Add(key)
-}
-
-// onNodeUpdate updates the Node Selector in the cache and queues the Node for processing.
-func (uc *upgradeController) onNodeUpdate(oldObj, newObj interface{}) {
-	oldNode := oldObj.(*v1.Node)
-	newNode := newObj.(*v1.Node)
-
-	// don't process resync or objects that are marked for deletion
-	if oldNode.ResourceVersion == newNode.ResourceVersion ||
-		!newNode.GetDeletionTimestamp().IsZero() {
-		return
-	}
-
-	key, err := cache.MetaNamespaceKeyFunc(newObj)
+// GetTopologyVersion polls the coordination points (Nodes and ConfigMaps) until
+// the master has reported a version
+func (uc *upgradeController) GetTopologyVersion(ctx context.Context) (int, error) {
+	ver := -1
+	err := wait.PollImmediateWithContext(ctx, 10*time.Second, 5*time.Minute, func(ctx context.Context) (bool, error) {
```

Comment:
why does GetTopologyVersion need to poll for 5 minutes? We didn't need that in the old version of the code.

Reply:
Good question. We have a very high assurance that the topology will already be reported, since we look for it only after the node's logical switch has been created. However, I decided to be paranoid. (The code, before this change, would just fail and restart if it couldn't determine the topo version).